lib/AST/Module.cpp

```diff
@@ -864,6 +864,23 @@ ProtocolConformanceRef ModuleDecl::lookupConformance(Type type,
   if (type->is<UnresolvedType>())
     return ProtocolConformanceRef(protocol);
 
+  // Tuples have builtin conformances implemented within the runtime.
+  // These conformances so far consist of Equatable.
+  if (auto tuple = type->getAs<TupleType>()) {
+    if (protocol == ctx.getProtocol(KnownProtocolKind::Equatable)) {
+      // Ensure that every element in this tuple conforms to Equatable.
+      for (auto eltTy : tuple->getElementTypes()) {
+        auto conformance = lookupConformance(eltTy, protocol);
```

Comment:
Sorry I missed this earlier, but we should really save the element conformances and use them as arguments to the builtin conformance. At runtime, we should similarly instantiate the witness table pattern with the collected witness tables for all of the element conformances. This will be more efficient, because we save having to dynamically look up those conformances, and will be robust in the case when there are multiple retroactive `Equatable`/`Comparable`/`Hashable` conformances for the same type in a process.

Reply:
I think I'm confused because you said earlier in this thread that it seems reasonable to have a single descriptor and witness table for all tuples. Is the instantiated witness table pattern separate from the witness table?

Also, when you mean arguments to the builtin conformance, are you suggesting arguments when we're instantiating the builtin conformance? (In hindsight this does make sense because they'll be readily accessible in SIL when we can devirtualize some witness calls.)

Reply:
Sorry for the confusion. We can still have a single descriptor, and the implementations of the protocol witness methods can remain constant, but the witness table itself would need to be instantiated with the conformances for all of the tuple elements, so that the implementation can call into them. For an analogy to what needs to happen, you could look at what happens when you declare a conditional conformance on a nominal type:

```
struct Pair<T, U> {}

extension Pair: Equatable where T: Equatable, U: Equatable { }

func foo<T: Equatable>(_: T.Type) {}

foo(Pair<Int, String>.self)
```

Because `Pair` itself does not require T or U to be Equatable, the type metadata does not have any record of the conformance. The `T: Equatable` and `U: Equatable` conformances will be collected and passed to instantiate the `Pair<T, U>` conformance when we call `foo`.